### PR TITLE
executor: expand args in build logs

### PR DESCRIPTION
**Pull request by flouthoc**
Expand user,builtin and heading args before logging build steps.

Closes: https://github.com/containers/buildah/issues/6327

#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test 
> /kind feature
> /kind flake
> /kind other

#### Does this PR introduce a user-facing change?

```release-note
executor: expand args in build logs
```

### CODE REVIEW

**Comment by flouthoc** [comment]
@vrothberg @nalind PTAL

**Review by vrothberg — COMMENTED** [review]
Changes and tests LGTM.

I didn't check, but maybe there's a similar behavior with ENV?

Thanks so much for the quick PR 🙏

File: imagebuildah/executor.go [hunk 1]
```diff
@@ -600,7 +600,19 @@ func (b *Executor) buildStage(ctx context.Context, cleanupStages map[int]*StageE
...
+			builtinArgs := argsMapToSlice(stage.Builder.BuiltinArgDefaults)
+			headingArgs := argsMapToSlice(stage.Builder.HeadingArgs)
+			userArgs := argsMapToSlice(stage.Builder.Args)
+			userArgs = append(builtinArgs, append(userArgs, headingArgs...)...)
+			argsCopy := args
+			for i := range args {
+				argsCopy[i], err = imagebuilder.ProcessWord(args[i].(string), userArgs)
+				if err != nil {
+					argsCopy[i] = args[i]
+					fmt.Printf("while replacing arg variables with values for format in logging %q: %w\n", args[i], err)
```
Comment (nalind) [inline imagebuildah/executor.go]: Why is this error message sent to stdout rather than one of the output destinations that's been specified for logging?

File: imagebuildah/executor.go [hunk 2]
```diff
@@ -600,7 +600,19 @@ func (b *Executor) buildStage(ctx context.Context, cleanupStages map[int]*StageE
 				prefix += ": "
 			}
 			suffix := "\n"
-			fmt.Fprintf(stageExecutor.executor.out, prefix+format+suffix, args...)
+			builtinArgs := argsMapToSlice(stage.Builder.BuiltinArgDefaults)
+			headingArgs := argsMapToSlice(stage.Builder.HeadingArgs)
+			userArgs := argsMapToSlice(stage.Builder.Args)
+			userArgs = append(builtinArgs, append(userArgs, headingArgs...)...)
+			argsCopy := args
```
Comment (nalind) [inline imagebuildah/executor.go]: If this is meant to create a copy of `args`, use `slices.Clone()`, otherwise they're both still pointing at the same array.

File: tests/bud.bats [hunk 3]
```diff
@@ -8844,3 +8844,42 @@ _EOF
...
+
+  # Test that unexpanded variables show as empty when no build-arg is provided
+  cat > $contextdir/Containerfile2 << _EOF
+FROM alpine
+ARG UNDEFINEDVAR
+RUN echo \$UNDEFINEDVAR
+_EOF
+
+  run_buildah build $WITH_POLICY_JSON -t ${target}-undefined -f $contextdir/Containerfile2 $contextdir
+  expect_output --substring "RUN echo "
```
Comment (nalind) [inline tests/bud.bats]: This would match whether a value was printed or not.

**Comment by flouthoc** [comment]
> This produces some weird corner cases, both with this patch and with `docker build`:
> 
> ```dockerfile
> FROM busybox
> ARG CMD=pwd
> RUN CMD=ls; echo $CMD
> ```

Similar behavior exists in `buildkit` as well I think this is bug in both `buildah` and `docker`. Should we fix it here ?

File: imagebuildah/executor.go [hunk 4]
```diff
@@ -600,7 +600,19 @@ func (b *Executor) buildStage(ctx context.Context, cleanupStages map[int]*StageE
 				prefix += ": "
 			}
 			suffix := "\n"
-			fmt.Fprintf(stageExecutor.executor.out, prefix+format+suffix, args...)
+			builtinArgs := argsMapToSlice(stage.Builder.BuiltinArgDefaults)
+			headingArgs := argsMapToSlice(stage.Builder.HeadingArgs)
+			userArgs := argsMapToSlice(stage.Builder.Args)
+			userArgs = append(builtinArgs, append(userArgs, headingArgs...)...)
+			argsCopy := slices.Clone(args)
```
Comment (Luap99) [inline imagebuildah/executor.go]: why is this cloning the full args array if we throw away all values, that is just causing unnecessary memory copies.
To preallocate use make([]string, 0, len(args)).

File: imagebuildah/executor.go [hunk 5]
```diff
@@ -600,7 +600,19 @@ func (b *Executor) buildStage(ctx context.Context, cleanupStages map[int]*StageE
...
+			for i := range args {
+				argsCopy[i], err = imagebuilder.ProcessWord(args[i].(string), userArgs)
```
Comment (Luap99) [inline imagebuildah/executor.go]: casting to string seems unsafe, the function signature accepts any, This means all non strings values are discarded basically.
If the log can only deal with string the function signature should be updated to only accept strings.

In principle I am also quite confused in how many different places ProcessWord() gets called, buildah seems to call this and replace values as one off in may palaces. I think it would be much cleaner to replace and and then keep the replaced values but I don't understand enough about the buildah design here.

File: imagebuildah/executor.go [hunk 6]
```diff
@@ -600,7 +600,19 @@ func (b *Executor) buildStage(ctx context.Context, cleanupStages map[int]*StageE
...
+				if err != nil {
+					argsCopy[i] = args[i]
+					fmt.Fprintf(stageExecutor.executor.err, "while replacing arg variables with values for format in logging %q: %w\n", args[i], err)
```
Comment (Luap99) [inline imagebuildah/executor.go]: you cannot use %w for normal formatting, only in fmt.Errorf()

File: imagebuildah/executor.go (hunk 5 above)
Comment (nalind) [reply imagebuildah/executor.go]: Am I looking at the right function?  `ProcessWord()` appears to take a string and a slice of strings, so type assertions shouldn't even be necessary.

Comment (Luap99) [reply imagebuildah/executor.go]: I am talking about the log function itself which defines args as `[]any` type. Everything that gets passed which is not a string is basically lost here due this type cast.

Comment (nalind) [reply imagebuildah/executor.go]: Ah, I follow now.  The assertion isn't checked, so I think that case would panic.  In practice, we only pass it strings.

Comment (flouthoc) [reply imagebuildah/executor.go]: I am processing only args with `string` type, I avoided changing function signature as it involves making changes in public API of some structs so will do that in a new PR. Practically all args are `strings` so this check is just preventing from `panic`.

File: imagebuildah/executor.go (hunk 6 above)
Comment (flouthoc) [reply imagebuildah/executor.go]: Done.

File: imagebuildah/executor.go (hunk 4 above)
Comment (flouthoc) [reply imagebuildah/executor.go]: Done.

**Comment by flouthoc** [comment]
> > > This produces some weird corner cases, both with this patch and with `docker build`:
> > > ```dockerfile
> > > FROM busybox
> > > ARG CMD=pwd
> > > RUN CMD=ls; echo $CMD
> > > ```
> > 
> > 
> > Similar behavior exists in `buildkit` as well I think this is bug in both `buildah` and `docker`. Should we fix it here ?
> 
> You don't event need build args to trigger this: `RUN test=1; echo ${test:-abc}`
> 
> And worse if you just have an unset non existing build arg buildah just removes from the log (uses empty string) which is also super confusing,
> 
> The issue states
> 
> > I would love Buildah to expand the variables as well. I couldn't wrap my head around an issue where the build succeeded but didn't provide the build artifact I expected. Trying docker build helped me debug the issue and correct the Containerfile.
> 
> Given that I would argue this change doesn't really improve the situation, it trades one problem for another. But if docker behaves like that and logs incorrectly then maybe matching its bugs makes sense? I don't personally care much either way.
> 
> But I think the correct thing is to only replace the vars that buildah actually replaced when executing the command which ties back to my comment about replacing the RUN string in to many different places.

I am keeping behavior same as buildkit for consistency and created an upstream issue with moby/buildkit#6139 to track it there.

File: imagebuildah/executor.go [hunk 7]
```diff
@@ -600,7 +607,26 @@ func (b *Executor) buildStage(ctx context.Context, cleanupStages map[int]*StageE
...
+						processedArg, err := imagebuilder.ProcessWord(arg, userArgs)
+						if err != nil {
+							processedArg = arg
+							fmt.Fprintf(stageExecutor.executor.err, "while replacing arg variables with values for format in logging %q: %v\n", arg, err)
+						}
+						processedArgs = append(processedArgs, processedArg)
+					} else {
+						processedArgs = append(processedArgs, arg)
+					}
+				}
```
Comment (Luap99) [inline imagebuildah/executor.go:627]: That still discards any arg which is not a string, the else block must be one lever out because it needs to keep the non string args otherwise the formatting string will be messed up badly.

File: imagebuildah/executor.go [hunk 8]
```diff
@@ -519,6 +520,12 @@ func (b *Executor) getImageTypeAndHistoryAndDiffIDs(ctx context.Context, imageID
 	return oci.OS, oci.Architecture, manifestFormat, oci.History, oci.RootFS.DiffIDs, nil
 }
 
+// Checks if is word is using any variable in the form of `$VAR`, `${VAR}` etc
+func containsVariable(arg string) bool {
+	re := regexp.MustCompile(`\$\{?[a-zA-Z_][a-zA-Z0-9_]*\}?`)
+	return re.MatchString(arg)
```
Comment (Luap99) [inline imagebuildah/executor.go:526]: That compiles the regex every single time which is quite inefficient. That should use something like regexp.Delayed() in the global scope.

But is the regex really right? At the very least it ignores escaping rules `\$var` is not a variable. ProcessWord() has already has a parser so this would need to match exactly what it is doing. I think it would be better to reuse the ProcessWord code and expose some function for this instead, yeah sure more complicated but the only sane way to avoid any differences between the var replacement code (ProcessWord) and this custom regex
